### PR TITLE
perf(distillation): defer teacher top-k Ray payloads

### LINKED ISSUE #1877: Ray data transfer bottleneck with long sequences/large batch size in distillation
When running distillation with long sequences, large batch size and topk settings (e.g., 32K response length, 512 batch size, 100 topk), Ray data transfer becomes a significant bottleneck, causing GPU underutilization.

**Affected Areas**

1. Training data dispatch (`policy_training/submit_training_futures`)
2. Teacher inference result return (`get_topk_logits` results)

**Observed Behavior**

With `max_total_sequence_length: 32768, train_global_batch_size: 512, topk_logits_k: 100`:

```other
Total step time: 945.93s

policy_training: 276.06s
├── submit_training_futures: 135.15s  ← should be fast, but takes 135s

teacher_logprob_inference: 200.66s
├── (GPU mostly idle during result transfer)

generation: 192.38s (38.8%)
```

`submit_training_futures` should be a quick async dispatch, but it takes **135 seconds** due to large tensor serialization overhead.

**Steps/Code to reproduce bug**

The config for reproducing the above bug:

```yaml
defaults: distillation_math_megatron.yaml

distillation:
    num_prompts_per_step: 512
    max_num_steps: 1000
    val_batch_size: 512
    val_period: 20
    val_at_start: false
    topk_logits_k: 100

loss_fn:
    kl_type: "reverse"

checkpointing:
    checkpoint_dir: "checkpoints/distillation-8b-test"
    save_period: 5

policy:
    model_name: "Qwen/Qwen3-8B"
    tokenizer:
        name: "Qwen/Qwen3-8B" 
    train_global_batch_size: 512
    generation_batch_size: 512
    max_total_sequence_length: 32768
    sequence_packing:
        enabled: true
    megatron_cfg:
        tensor_model_parallel_size: 2
        expert_tensor_parallel_size: 1
        expert_model_parallel_size: 1
        pipeline_model_parallel_size: 1
        context_parallel_size: 4
        sequence_parallel: true
    generation:
      vllm_cfg:
          tensor_parallel_size: 4
          expert_parallel_size: 1
    make_sequence_length_divisible_by: 64 
teacher:
    model_name: "Qwen/Qwen3-8B"
    tokenizer:
        name: "Qwen/Qwen3-8B"
    max_total_sequence_length: 32768
    sequence_packing:
        enabled: true 
    megatron_cfg:
        tensor_model_parallel_size: 2
        expert_tensor_parallel_size: 1
        expert_model_parallel_size: 1
        pipeline_model_parallel_size: 1
        context_parallel_size: 4  
        sequence_parallel: true   
    make_sequence_length_divisible_by: 64 

```

**Expected behavior**

Optimize the large data transfer.

**Additional context**

### PR DESCRIPTION
**Pull request by yuchenwang3**
# What does this PR do?

Keep DP-local teacher top-k payloads in the Ray object store and pass their references directly to matching student workers instead of materializing the large tensors in the driver.

The fast path is enabled only when teacher and student have compatible static DP layouts. Dynamic batching, mismatched DP sizes, incompatible sequence-packing layouts, and partial batches continue to use the existing materialized path.

Worker methods return a small wrapper containing the worker-local payload reference. The driver still materializes every wrapper, so failures from replicated workers propagate before representative payload references are selected. Ray sharding also preserves existing ObjectRefs rather than wrapping them in another object-store entry.

# Issues

Closes #1877

# Usage

No user-facing configuration change is required. Compatible distillation runs select the deferred transfer path automatically.

# Validation

Exact source under test:

- Base: `71a08314bc3257a15ef04b507309649e3d3f0f68`
- Patch: `20cfa59cc405067629d1c99562f10687f396bfb3`
- Patch tree: `036dd95966651201da0998d009f939dbec630815`
- Source clean before and after testing

Focused tests:

- `tests/unit/models/policy/test_deferred_topk.py`: 8 passed, 0 skipped
- `tests/unit/distributed/test_worker_groups.py`: 28 passed, 1 unrelated Nsight environment skip
- `tests/unit/algorithms/test_distillation.py`: 23 passed, 0 skipped
- The ObjectRef preservation and non-representative worker error-propagation tests both executed and passed

Controlled 8×B200 Ray transfer benchmark:

| Payload | Repeats | Base median / p95 | Patch median / p95 | Speedup | Base / Patch driver RSS delta |
| --- | ---: | ---: | ---: | ---: | ---: |
| 0.25 GB | 3 | 1.018 / 1.037 s | 0.229 / 0.329 s | 4.44× | 752.9 MB / 102 KB |
| 1 GB | 3 | 4.194 / 4.215 s | 0.929 / 1.212 s | 4.51× | 2.75 GB / 98 KB |
| 4 GB | 3 | 17.644 / 17.947 s | 3.697 / 3.857 s | 4.77× | 11.00 GB / 94 KB |
| 19.7 GB | 2 | 90.927 / 90.954 s | 17.115 / 17.121 s | 5.31× | 54.18 GB / 303 KB |

Every measured run used 8 actors, completed all repeats with exit 0, matched checksums, resolved ObjectRefs automatically, and reported zero object spilling. The final-patch harness models the production path: worker-local `ray.put`, small wrapper materialization in the driver, then direct payload-reference handoff. A separate probe confirmed that an error from a non-representative worker still fails the driver.

This is a controlled transfer benchmark for the bottleneck mechanism rather than a claim about full end-to-end training-step speedup.

Local checks also passed: Ruff lint, Ruff format check, Python compileall, and `git diff --check`.

# Before your PR is "Ready for review"

- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? No documentation change is needed because this is an automatic internal transfer optimization with no new user-facing configuration.

# Additional Information

The commit is signed off under DCO by Yuchen Wang.

### CODE REVIEW

**Comment by yuchenwang3** [comment]
@nvidia-nemo/rl_reviewers_algo @nvidia-nemo/rl_reviewers_infra — final validation is complete on head `20cfa59cc405067629d1c99562f10687f396bfb3`: the 8×B200 controlled transfer matrix passed at all four payload sizes with matching checksums, zero spill, and 4.44–5.31× speedup. Focused suites passed 8 + 28 + 23 tests; the only skip was an unrelated Nsight environment check. Could one of you route the review, add `CI:L1`, and approve `/ok to test 20cfa59cc405067629d1c99562f10687f396bfb3`? Thanks.

**Comment by yuchenwang3** [comment]
@zpqiu, this PR addresses the teacher-result transfer bottleneck you reported in #1877 by deferring compatible DP-local top-k payloads through Ray object references. The PR documents controlled transfer measurements and fallback coverage; those measurements are not an end-to-end training speedup claim. Could you take a look at the approach or help identify the appropriate distillation reviewer? Maintainer CI authorization is also still needed. Thanks.

**Comment by yuchenwang3** [comment]
Synced with current main and resolved the policy-worker conflict, retaining both deferred top-k transfer and the new pre-refit synchronization hook. No change to the transfer algorithm.

Changed-file Ruff/format checks pass. Local isolated CPU checks confirm the deferred methods are unchanged and the upstream refit hook is preserved. The earlier Ray/GPU measurements remain results for the previously documented revision; I have not rerun that benchmark on this merge. Upstream runner validation still needs maintainer authorization.

**Comment by yuchenwang3** [comment]
@ananthsub Rebased at `eff2617084`, keeping the deferred top-k transfer logic and upstream pre-refit sync hook. Could you review the approach for #1877 or help route it? Local lint passes; the earlier Ray/GPU measurements haven't been rerun on this head.

**Comment by zpqiu** [comment]
Thank you for this contribution! Could we evaluate reusing the existing data-plane/TQ path for teacher top-k payloads? `get_teacher_logprobs_presharded()` already writes full teacher payloads directly to the data plane. I think @ZhiyuLi-Nvidia  @zyzhou5  might have some suggestions on the TQ path.

**Comment by yuchenwang3** [comment]
Thanks @zpqiu, I traced the TQ path and agree it looks suitable for top-k too. The existing `pack_jagged_fields()` / `materialize()` codec handled `[N, S, K]` payloads in four local CPU checks: BF16/FP32 logits, int64 indices, unequal sequence lengths, and reordered rows. That probe explicitly marks the two fields token-aligned; it doesn't exercise a TQ backend or training.

The remaining wiring would be:

- Register `teacher_topk_logits` / `teacher_topk_indices` in the partition and token-aligned field registry.
- Add a top-k presharded entrypoint using the existing fetch, globally planned packing, and worker write-back helpers. [`get_teacher_logprobs_presharded()`](https://github.com/NVIDIA-NeMo/RL/blob/eff2617084fc12b1054789e1cabaccf702c75f62/nemo_rl/data_plane/worker_mixin.py#L964-L1045) currently dispatches token logprobs or full hidden-state/logit payloads.
- Reuse [`TQPolicy.train_from_meta(train_fields=...)`](https://github.com/NVIDIA-NeMo/RL/blob/eff2617084fc12b1054789e1cabaccf702c75f62/nemo_rl/models/policy/tq_policy.py#L365-L432) with distillation's field set, and connect partition setup/cleanup in the trainer. Distillation currently constructs plain `Policy` instances and passes `BatchedDataDict` batches.

I'd favor reusing those components over maintaining another transport. Would you prefer that I migrate distillation's step lifecycle to TQ in this PR and remove the deferred-Ray path? @ZhiyuLi-Nvidia @zyzhou5, guidance on that scope would help before I rework the driver.
